testing/sev_snp_hello_world_kernel/src/ghcb.rs

```diff
@@ -0,0 +1,120 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use core::mem::MaybeUninit;
+use sev_guest::ghcb::{Ghcb, GhcbProtocol};
+use x86_64::{
+    addr::{PhysAddr, VirtAddr},
+    instructions::tlb::flush_all,
+    registers::control::Cr3,
+    structures::paging::page_table::{PageTable, PageTableFlags},
+};
+
+/// The mask for the encrypted bit. For now we assume that we will be running on AMD Arcadia-Milan
+/// CPUs, which use bit 51.
+const ENCRYPTED_BIT: u64 = 1 << 51;
+
+#[link_section = ".ghcb"]
+static mut GHCB: MaybeUninit<Ghcb> = MaybeUninit::uninit();
+
+/// Initializes the GHCB and shares it with the hypervisor.
+pub fn init_ghcb(snp_enabled: bool) -> GhcbProtocol<'static> {
+    // Safety: this data structure is placed in valid memory so we won't page fault when accessing
+    // it. We reset the value of the GHCB immediately after shareing it with the hypervisor, so it
+    // will be fine if it is not initialized.
+    let ghcb: &mut Ghcb = unsafe { GHCB.assume_init_mut() };
+    share_ghcb_with_hypervisor(ghcb, snp_enabled);
+    ghcb.reset();
+    GhcbProtocol::new(ghcb)
+}
+
+/// Marks the page containing the GHCB data structure to be shared with the hypervisor.
+fn share_ghcb_with_hypervisor(ghcb: &Ghcb, snp_enabled: bool) {
+    let ghcb_address = VirtAddr::new(ghcb as *const Ghcb as usize as u64);
+    if snp_enabled {
+        // This code will eventually also mark the page as shared in the RMP and use the MSR
+        // protocol to register the page with the hypervisor.
+        unimplemented!("SNP support is not yet implemented.");
+    }
+    // Mark the page as not encrypted.
+    set_encrypted_bit_for_page(&ghcb_address, false);
+}
+
+/// Sets the encrypted bit for the page that contains the address.
+///
+/// We assume for now that we will always use 2MiB huge pages.
+fn set_encrypted_bit_for_page(address: &VirtAddr, encrypted: bool) {
```

Comment:
IIUC this function is a page table walk for finding a specific phys address.
Is this page table walk used by other parts of the kernel? Could it be shared somewhere, so we won't have code duplication?

Reply:
Yes, this does a manual page table walk because we need to find the right entry to modify. When we add more advanced page table manipulation features to the main restricted kernel we could change this code to reuse a subset of that logic, but for now I don't think there is really a suitable library where this could go.

Reply:
There are page table wrappers in the `x86_64` library, but those don't know anything about the encrypted bit, so probably will not be useful here. We will likely need to implement our own page table abstractions at one point.